### PR TITLE
Active task serialization

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**
Added serialization objects for the active steps that are currently included in SageResearch-iOS. This also includes a bunch of code cleanup and fixing warnings. 

Note: This does *not* include setting up permission handling and async actions in the navigator.

### CODE REVIEW

File: assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ * The recorder types are split into two categories of [raw] sensor data and [calculated] values. The [calculated]
+ * values are are black box values calculated by the platform libraries and may vary between devices.
+ *
+ * - Raw data:
+ *   - [Accelerometer]: Raw accelerometer reading.
+ *   - [Gyro]: Raw gyroscope reading.
+ *   - [Magnetometer]: Raw magnetometer reading.
+ *
+ * - Calculated values:
+ *   - [Attitude]: Calculated orientation of the device using the gyro and magnetometer (if appropriate).
```
Comment (Erin-Mounts) [inline assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt:52]:
```suggestion
 *   - [Attitude]: Calculated orientation of the device using the motion sensors and magnetometer (if appropriate).
```

File: assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            UserAcceleration
+        )
+    }
+}
+
+/**
+ * A [MotionRecord] is a serializable implementation of a [SampleRecord] that can be used to record a sample from one
+ * of the motion sensors or calculated vectors. All properties on the record ard defined as `null` to allow using this
+ * same serialization object to define both path markers that are added to the data stream to mark step transitions
+ * and the results of the sensor.
```
Comment (Erin-Mounts) [inline assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt:94]: Suggested for the last sentence:

Path markers are added to the data stream to mark step transitions using this same serialization object by setting all properties on the record to `null`.

File: assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+ * sensors. This data structure allows for defining all records within a single file.
+ *
+ * Vector properties:
+ * - [x]: The `x` component of the vector measurement for this sensor sample.
+ * - [y]: The `y` component of the vector measurement for this sensor sample.
+ * - [z]: The `z` component of the vector measurement for this sensor sample.
+ * - [w]: The `w` component of the vector measurement for this sensor sample. (Used by the attitude quaternion.)
+ *
+ * Additional properties:
+ * - [eventAccuracy]: A number marking the sensor accuracy of the magnetic field sensor.
```
Comment (Erin-Mounts) [inline assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt:106]: Should this have a different name? Or does it actually indicate the accuracy of the sample regardless of which sensor and not just for the magnetometer?

File: assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      * the horizontal plane.
+      */
+     @SerialName("Z-Up") XArbitraryZVertical,
+
+     /**
+      * Describes a reference frame in which the Z axis is vertical and the X axis points toward magnetic north.
+      *
+      * - note: Using this reference frame may require user interaction to calibrate the magnetometer.
+      */
+     @SerialName("North-West-Up") XMagneticNorthZVertical;
```
Comment (Erin-Mounts) [inline assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt:141]: I expect for most use cases (exception: those involving studies of human sensitivity to Earth's magnetic field) that true north would be more useful, if heading is needed at all. If we're including this, should we include the true north reference frame as an option as well? (would suggest just replacing this one, but it does require location to be available as well as a calibrated magentometer.)

File: assessmentModel/src/commonMain/kotlin/serialization/InputItems.kt [hunk 5]
```diff
@@ -38,8 +38,7 @@ val inputItemSerializersModule = SerializersModule {
 }
 
 /**
- * A [InputItemObject] is intended to implement shared code for serialization of the simple data types. This will
- * work to deserialize some of the existing input items that use the [DataType] to define their type.
+ * A [InputItemObject] is intended to implement shared code for serialization of the simple data types.
```
Comment (Erin-Mounts) [inline assessmentModel/src/commonMain/kotlin/serialization/InputItems.kt:41]:
```suggestion
 * An [InputItemObject] is intended to implement shared code for serialization of the simple data types.
```

File: assessmentModel/src/commonMain/kotlin/serialization/InputItems.kt [hunk 6]
```diff
@@ -228,18 +227,18 @@ data class ChoiceItemWrapper(val choice: ChoiceOption,
...
-    Height, Weight, BloodPressure;
-}
-
-/**
- * The measurement range is used to determine units that are appropriate to the size of the person.
- */
-enum class MeasurementRange : StringEnum {
-    Adult, Child, Infant;
-}
+///**
```
Comment (Erin-Mounts) [inline assessmentModel/src/commonMain/kotlin/serialization/InputItems.kt:230]: TODO or delete

File: assessmentModel/src/commonMain/kotlin/AsyncActionConfiguration.kt [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+     * Should the file used in a previous run of a recording be deleted or should a unique file name be used to
+     * identify the file when a step is repeated? Default = `true`.
+     */
+    val shouldDeletePrevious: Boolean
+        get() = true
+
+    /**
+     * Set the flag to `true` to encode the samples as a CSV file. Default = `false`.
+     */
+    val usesCSVEncoding : Boolean
```
Comment (nategbrown9) [inline assessmentModel/src/commonMain/kotlin/AsyncActionConfiguration.kt:61]: Are all recorders going to have the ability to encode samples as a CSV file? Wondering if this should be in a separate interface or just on the implementation classes.

File: assessmentModel/src/commonMain/kotlin/common.kt [hunk 8]
```diff
@@ -11,4 +11,13 @@ expect object UUIDGenerator {
...
+}
+
+expect class Product {
+    val user: String
+}
+
+expect object Factory {
+    fun create(config: Map<String, String>): Product
+    val platform: String
+}
```
Comment (nategbrown9) [inline assessmentModel/src/commonMain/kotlin/common.kt:23]: Don't think we need this code.

File: assessmentModel/src/commonMain/kotlin/AsyncActionConfiguration.kt (hunk 7 above)
Comment (syoung-smallwisdom) [reply assessmentModel/src/commonMain/kotlin/AsyncActionConfiguration.kt:61]: Well, they all do in SageResearch b/c it's easy but yeah, good point.

File: assessmentModel/src/commonMain/kotlin/AsyncActionConfiguration.kt [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     * Whether or not the recorder must be allowed to run when the app is in the background. Default = `false`.
+     */
+    val requiresBackground: Boolean
+        get() = false
+
+    /**
+     * Should the file used in a previous run of a recording be deleted or should a unique file name be used to
+     * identify the file when a step is repeated? Default = `true`.
```
[the remark this answers is not in the document]
Comment (syoung-smallwisdom) [reply assessmentModel/src/commonMain/kotlin/AsyncActionConfiguration.kt:53]: This is correctly phrased as a question

File: assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt (hunk 3 above)
Comment (syoung-smallwisdom) [reply assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt:106]: We have years of code files with the existing name for the field so I'm gonna go with "no".

File: assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt (hunk 4 above)
Comment (syoung-smallwisdom) [reply assessmentModel/src/commonMain/kotlin/recorders/MotionRecorderConfiguration.kt:141]: Copied directly from the Apple documentation on the sensor.

File: assessmentModel/src/commonMain/kotlin/serialization/InputItems.kt (hunk 6 above)
Comment (syoung-smallwisdom) [reply assessmentModel/src/commonMain/kotlin/serialization/InputItems.kt:230]: There is already a TODO comment

File: assessmentModel/src/commonMain/kotlin/common.kt (hunk 8 above)
Comment (syoung-smallwisdom) [reply assessmentModel/src/commonMain/kotlin/common.kt:23]: Still referenced in the Android main activity so I decided against deleting it.

**Pull request merged**